tlatools/org.lamport.tlatools/test/tla2sany/semantic/SemanticCorpusTests.java

```diff
@@ -0,0 +1,279 @@
+ *    accepting an expression (which could be an operator reference or label)
+ *    along with an assertion about its level.
+ *
+ * Both RefersTo and IsLevel are defined in a module Semantics.tla which is
+ * included in the test corpus directory and resolved during parsing. The
+ * test module is then searched to find all assertions, then their parameters
+ * are analyzed to see whether the assertions are correct.
+ */
+@RunWith(Parameterized.class)
+public class SemanticCorpusTests {
+
+  private static class SemanticTestCase {
+    public final Path ModulePath;
+    public SemanticTestCase(Path modulePath) {
+      this.ModulePath = modulePath;
+    }
+    // Used by JUnit to identify this test case.
+    public String toString() {
+      return this.ModulePath.getFileName().toString();
+    }
+  }
+
+  private static class SemanticsModule {
+    public final ExternalModuleTable Modules;
+    public final OpDefNode RefersTo;
+    public final OpDefNode IsLevel;
+    public final OpDefNode ConstantLevel;
+    public final OpDefNode VariableLevel;
+    public final OpDefNode ActionLevel;
+    public final OpDefNode TemporalLevel;
+    public SemanticsModule(ExternalModuleTable modules) {
+      this.Modules = modules;
+      ModuleNode semanticsModule = modules.getModuleNode("Semantics");
+      this.RefersTo = semanticsModule.getOpDef("RefersTo");
+      this.IsLevel = semanticsModule.getOpDef("IsLevel");
+      this.ConstantLevel = semanticsModule.getOpDef("ConstantLevel");
+      this.VariableLevel = semanticsModule.getOpDef("VariableLevel");
+      this.ActionLevel = semanticsModule.getOpDef("ActionLevel");
+      this.TemporalLevel = semanticsModule.getOpDef("TemporalLevel");
+    }
+    public int getExpectedLevel(final SymbolNode level) {
+      if (this.ConstantLevel == level) {
+        return LevelConstants.ConstantLevel;
+      } else if (this.VariableLevel == level) {
+        return LevelConstants.VariableLevel;
+      } else if (this.ActionLevel == level) {
+        return LevelConstants.ActionLevel;
+      } else if (this.TemporalLevel == level) {
+        return LevelConstants.TemporalLevel;
+      } else {
+        throw new IllegalArgumentException("Not a level: " + level.getName());
+      }
+    }
+    public List<OpApplNode> findRefersToAssertions() {
+      return findAllReferences(this.Modules.getRootModule(), this.RefersTo);
+    }
+    public List<OpApplNode> findIsLevelAssertions() {
+      return findAllReferences(this.Modules.getRootModule(), this.IsLevel);
+    }
+  }
+
+  /**
+   * The directory containing all the semantic corpus test files.
+   */
+  private static final String CORPUS_DIR = "test/tla2sany/semantic/corpus";
+
+  /**
+   * Finds all semantic corpus test files.
+   */
+  @Parameters(name = "{index}: {0}")
+  public static List<SemanticTestCase> getTestFiles() throws IOException {
+    Path corpusDir = Paths.get(CommonTestCase.BASE_DIR).resolve(CORPUS_DIR);
```

Comment:
If the tests are executed on the tla2tools.jar file, loading the .tla files with `Files` will fail. Since the .tla files are on the classpath, it’s better to use `classloader.getResourceAsStream(…)` instead.

Reply:
I think this sort of change might have to be delayed until the API prototyped in #1125 is integrated with existing resource loading code.

Reply:
Please add a code-level TODO.